src/iop/mlens.c

```diff
@@ -0,0 +1,452 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2010-2020 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include "bauhaus/bauhaus.h"
+#include "common/interpolation.h"
+#include "develop/imageop.h"
+#include "develop/imageop_gui.h"
+#include "develop/tiling.h"
+#include "gui/gtk.h"
+#include "iop/iop_api.h"
+
+#include <gtk/gtk.h>
+#include <stdlib.h>
+
+#define NKNOTS 16
+
+DT_MODULE_INTROSPECTION(1, dt_iop_mlens_params_t)
+
+typedef struct dt_iop_mlens_params_t
+{
+  gboolean cor_dist;  // $DEFAULT: 1 $DESCRIPTION: "correct distortion"
+  gboolean cor_ca;    // $DEFAULT: 1 $DESCRIPTION: "correct chromatic aberration"
+  gboolean cor_vig;   // $DEFAULT: 1 $DESCRIPTION: "correct vignetting"
+  float cor_dist_ft;  // $DEFAULT: 1 $MIN: 0 $MAX: 2 $DESCRIPTION: "distortion fine-tune"
+  float cor_vig_ft;   // $DEFAULT: 1 $MIN: 0 $MAX: 2 $DESCRIPTION: "vignette fine-tune"
+  float scale;
+} dt_iop_mlens_params_t;
+
+typedef struct dt_iop_mlens_gui_data_t
+{
+  GtkWidget *cor_dist, *cor_ca, *cor_vig, *cor_dist_ft, *cor_vig_ft;
+} dt_iop_mlens_gui_data_t;
+
+const char *name()
+{
+  return _("manufacturer lens correction");
+}
+
+const char *aliases()
+{
+  return _("vignette|chromatic aberrations|distortion");
+}
+
+const char *description(dt_iop_module_t *self)
+{
+  return dt_iop_set_description(self, _("correct lenses optical flaws"),
+                                _("corrective"),
+                                _("linear, RGB, scene-referred"),
+                                _("geometric and reconstruction, RGB"),
+                                _("linear, RGB, scene-referred"));
+}
+
+int default_group()
+{
+  return IOP_GROUP_CORRECT | IOP_GROUP_TECHNICAL;
+}
+
+int operation_tags()
+{
+  return IOP_TAG_DISTORT;
+}
+
+int flags()
+{
+  return IOP_FLAGS_ALLOW_TILING | IOP_FLAGS_TILING_FULL_ROI | IOP_FLAGS_ONE_INSTANCE | IOP_FLAGS_UNSAFE_COPY;
+}
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+static int init_coeffs(const dt_image_t *img, const dt_iop_mlens_params_t *d,
```

Comment:
Looking at this in more detail:
Why don't we do this processing on import and store a manufacturer independent matrix for distortion, ca and vignetting? That would simplify the code here, keep dt_image_t small and simply the caching code.

Reply:
As the models are reverse engineered there is no guarantee that the coefficient processing is correct (I think it is, but I am not sure).  Hence why I placed it in a function.  Incidentally, this is why I am against the idea of caching these coefficients in any manner at all.  It is almost certainly more trouble than it's worth.  When you're opening a 20 MiB raw file and doing highly intensive processing on it, shaving off a few milliseconds by caching metadata is basically a rounding error.  Moreover, it leads to issues such as those you've outlined above regarding how to deal with files whose metadata was extracted before support was added.

Reply:
Fair point.